pkg/cluster/minio.go

```diff
@@ -0,0 +1,373 @@
+	err = c.Config.Client.AppsV1().Deployments(ns).Delete(ctx, "minio", metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to clean up minio deployment: %w", err)
+	}
+	err = c.Config.Client.CoreV1().PersistentVolumeClaims(ns).Delete(ctx, "minio-pv-claim", metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to clean up minio pvc: %w", err)
+	}
+	backgroundProp := metav1.DeletePropagationBackground
+	err = c.Config.Client.BatchV1().Jobs(ns).Delete(ctx, "migrate-minio-ha", metav1.DeleteOptions{PropagationPolicy: &backgroundProp})
+	if err != nil {
+		return fmt.Errorf("failed to clean up minio migration job: %w", err)
+	}
+
+	return nil
+}
+
+func (c *Controller) waitForJobCompletion(ctx context.Context, jobName string, jobNS string) error {
+	for {
+		runningJob, err := c.Config.Client.BatchV1().Jobs(jobNS).Get(ctx, jobName, metav1.GetOptions{})
+		if err != nil {
+			return fmt.Errorf("failed to get job %s in %s: %w", jobName, jobNS, err)
+		}
+		if runningJob.Status.Succeeded > 0 {
+			return nil
+		}
+		time.Sleep(time.Second * 5)
+	}
+}
+
+// MaybeRebalanceMinioServers
+// first, check if minio is healthy. If it is not healthy, don't do anything.
+// (this may require manual intervention, or may be resolved when nodes come back online)
+// Then, check for pods that have been unschedulable for more than 5 minutes
+// and delete the underlying volume + pod.
+// This will allow the pod to be scheduled on a node that still exists and for data to be rebalanced there.
+// After ensuring that there are as many running replicas as possible, we can rearrange replicas to maximize the number
+// of nodes that can be lost before losing data. To do this, first
+// check if more than ceil(replicas/nodes) replicas exist on one node.
+// If it does, see if we can reschedule one of those replicas safely.
+// If such a node does not exist, instead look for nodes with less than floor(replicas/nodes) replicas,
+// and if it exists reschedule a replica from a node with ceil(replicas/nodes).
+func (c *Controller) MaybeRebalanceMinioServers(ctx context.Context, ns string) error {
+	if !c.haMinioHealthy(ns) {
+		c.Log.Debugf("Not rebalancing Minio pods as the statefulset is not healthy")
+		return nil
+	}
+
+	minioPods, err := c.Config.Client.CoreV1().Pods(ns).List(ctx, metav1.ListOptions{LabelSelector: "app=ha-minio"})
+	if err != nil {
+		return fmt.Errorf("faled to get ha-minio pods: %w", err)
+	}
+
+	var unschedulablePod *corev1.Pod
+	for idx, minioPod := range minioPods.Items {
+		if minioPod.Status.Phase != corev1.PodPending {
+			continue
+		}
+		for _, condition := range minioPod.Status.Conditions {
+			if condition.Type == corev1.PodScheduled && condition.Status == corev1.ConditionFalse {
+				if time.Since(condition.LastTransitionTime.Time) > time.Minute*5 {
+					// pod has been unschedulable for 5 minutes, we can assume it should be rescheduled
+					unschedulablePod = &minioPods.Items[idx]
```

Comment:
break?

Reply:
I'll actually just make it return from inside the loop